File: swift/graph/adjacencyList.swift
// GraphNode used for adjacency list
class GraphNode {
  var val: String
  var neighbors: [String]

  init(_ val: String) {
    self.val = val
    neighbors = []
  }
}

// Or use a Hashmap
let adjList = ["A": [], "B": []]

// Given directed edges, build an adjacency list
func buildAdjList() -> [String: [String]] {
  let edges = [["A", "B"], ["B", "C"], ["B", "E"], ["C", "E"], ["E", "D"]]
  var adjList = [String: [String]]()
  
  for edge in edges {
    let src = edge[0], dst = edge[1]
    if adjList[src] == nil {
      adjList[src] = []
    }
    if adjList[dst] == nil {
      adjList[dst] = []
    }
    adjList[src]!.append(dst)
  }
  return adjList
}

// Count paths (backtracking)
func dfs(_ node: String, _ target: String, _ adjList: [String: [String]], _ visit: inout Set<String>) -> Int {
  if visit.contains(node) {
    return 0
  }
  if node == target {
    return 1
  }

  var count = 0
  visit.insert(node)
  for neighbor in adjList[node]! {
    count += dfs(neighbor, target, adjList, &visit)
  }
  visit.remove(node)
  return count
}

// Shortest path from node to target
func bfs(_ node: String, _ target: String, _ adjList: [String: [String]]) -> Int {
  var length = 0
  var visit = Set<String>()
  visit.insert(node)
  var queue: [String] = []
  queue.append(node)

  while queue.count > 0 {
    for _ in 0..<queue.count {
      let curr = queue.removeFirst()
      if curr == target {
        return length
      }

      for neighbor in adjList[curr]! {
        if !visit.contains(neighbor) {
          visit.insert(neighbor)
          queue.append(neighbor)
        }
      }
    }
    length += 1
  }
  return length
}

File: csharp/graph/AdjacencyList.cs
 
    public class AdjacencyList
    {

        public Dictionary<String, List<String>> BuildAdjList()
        {
            Dictionary<String, List<String>> adjList = new();

            string[][] edges = { new[] { "A", "B" }, new[] { "B", "C" } };
            HashSet<String> visit = new();

            adjList.Add("A", new List<String>());
            adjList.Add("B", new List<String>());

            foreach (var edge in edges)
            {
                String src = edge[0], dst = edge[1];
                if (!adjList.ContainsKey(src))
                {
                    adjList.Add(src, new List<String>());
                }
                if (!adjList.ContainsKey(dst))
                {
                    adjList.Add(dst, new List<String>());
                }
                adjList[src].Add(dst);
            }
            return adjList;
        }

        // Count paths (backtracking)
        public int Dfs(String node, String target, Dictionary<String, List<String>> adjList, HashSet<String> visit)
        {
            if (visit.Contains(node))
            {
                return 0;
            }
            if (node == target)
            {
                return 1;
            }
            int count = 0;
            visit = new HashSet<String>
            {
                node
            };
            foreach (String neighborin in adjList[node])
            {
                count += Dfs(neighborin, target, adjList, visit);
            }
            visit.Remove(node);
            return count;
        }

        // Shortest path from node to target.
        public int BFS(String node, String target, Dictionary<String, List<String>> adjList)
        {
            int length = 0;
            HashSet<String> visit = new HashSet<String>();
            Queue<String> q = new Queue<string>();
            visit.Add(node);
            q.Enqueue(node);

            while (q.Count != 0)
            {
                int queueLength = q.Count;
                for (int i = 0; i < queueLength; i++)
                {
                    String curr = q.Peek();
                    q.Dequeue();
                    if (curr.Equals(target))
                    {
                        return length;
                    }
                    foreach (String neighbor in adjList[curr])
                    {
                        if (!visit.Contains(neighbor))
                        {
                            visit.Add(neighbor);
                            q.Enqueue(neighbor);
                        }
                    }
                }
                length++;
            }
            return length;
        }
    }